exchange.c

```diff
@@ -496,3 +506,31 @@ static int domain_countToGo(ptrdiff_t nlimit, int (*layoutfunc)(int p), int fail
     }
     return ret;
 }
+
+static void
+realloc_secondary_data(int newMaxPartSph, int newMaxPartBh, int newMaxPartStar)
+{
+    size_t bytes = newMaxPartSph * sizeof(struct sph_particle_data) +
+        newMaxPartStar * sizeof(struct star_particle_data) + newMaxPartBh * sizeof(struct bh_particle_data);
+    SphP = myrealloc(SphP, bytes);
```

Comment:
It feels a bit uncomfortable with these memmove's and shifting of positions. 
Remember that we only use a PI to locate the secondary data. So one possibility to avoid memcopy is bucketing:

1. keep a linked list of allocations
2. when resolving PI, go though the buckets and till the correct one, find the relative offset in the bucket and return the ptr.

When there are too many buckets the code will slow down badly; a warning shall be issued. The main point of the growing mechanism is to ensure that it doesn't crash out-of-slots right away.

Reply:
I bet you any money that the memmove is faster.

Note that the BH is only moved when SPH gets larger (I could add a (if(src!=dest) memmove(dest,src), but I'm assuming that any competent memmove implementation does that itself anyway) and the initial allocation for Sph is the same as it always was - since we have only once had a situation where Sph ran out of memory, I expect that in practice the BH will never move.

So really it is just the stars, which are very small - sizeof(StarP[0]) is 32 bytes, I think. We are talking memmoving a total of ~30MB. memmove is very fast. Also they get moved max once or twice a simulation. Whereas the bucket access penalty will be every timestep during GC.